src/common/configurator.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2009-2018 Toby Schneider (http://gobysoft.org/index.wt/people/toby)
+//                     GobySoft, LLC (2013-)
+//                     Massachusetts Institute of Technology (2007-2014)
+//                     Community contributors (see AUTHORS file)
+//
+//
+// This file is part of the Goby Underwater Autonomy Project Libraries
+// ("The Goby Libraries").
+//
+// The Goby Libraries are free software: you can redistribute them and/or modify
+// them under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or
+// (at your option) any later version.
+//
+// The Goby Libraries are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Goby.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef CONFIGURATOR_20181205H
+#define CONFIGURATOR_20181205H
+
+#include "goby/common/configuration_reader.h"
+#include "goby/common/core_helpers.h"
+#include "goby/common/protobuf/app3.pb.h"
+
+namespace goby
+{
+namespace common
+{
+template <typename Config> class ConfiguratorInterface
+{
+  public:
+    void finalize()
+    {
+        if (!config_finalized_)
```

Comment:
In what case would this be called twice? Can I suggest reducing the cyclomatic complexity with just

```c++
if (config_finalized_)
    return;
```

Unless there is a reason why the top part should only execute once but the `debug_cfg()` part would want to check twice (and fail the first time)?

Or, since it smells like bad code if `finalize()` gets called twice, maybe throw an exception if it does.

Reply:
No longer relevant: Everything is done in the constructor of ConfiguratorInterface and its subclasses.